Source/bplaysrc/bplay.1
.TH BPLAY 1 "10 June 1994"
.SH NAME
bplay, brec \- buffered sound recording/playing
.SH SYNOPSIS
.B bplay
[\-S] [\-s speed] [\-b bits] [[\-t secs] | [\-T samples]]
[[\-j secs] | [\-J samples]] [file]
.sp
.B brec
[\-S] [\-s speed] [\-b bits] [[\-t secs] | [\-T samples]] [file]
.SH DESCRIPTION
.B bplay
copies data from the named sound file (or the standard input if no filename
is given) to the audio device.
.PP
.B brec
copies data from the audio device to the named sound file (or the
standard output if no filename is present).
.PP
These programs are eventually intended to be drop\-in replacements for the
.B vplay
and
.B vrec
programs by Michael Beck (beck@informatik.hu-berlin.de).
.SS OPTIONS
.TP
.I "\-S"
Sound file is stereo.
.TP
.I "\-s speed"
The speed in samples per second.
.TP
.I "\-b bits"
The number of bits per sample. Only 8 and 16 are currently supported.
.TP
.I "\-t secs"
The number of seconds to be played or recorded.
.TP
.I "\-T samples"
The number of samples to be played or recorded.
.TP
.I "\-j secs"
When playing, the number of seconds to skip at the beginning of the input before
playing.
.TP
.I "\-J samples"
When playing, the number of samples to skip at the beginning of the input before
playing.
.SH FILES
.I /dev/dsp
The audio device.
.SH BUGS
.PP
The \-t, \-T, \-j and \-J options may do strange things when playing VOC
files.
.PP
There are limitations on recording VOC format files - specifically
VOC files are only
recorded in the newer 1.20 version of the format, which older player
programs may choke on. These limitations should be fixed in v1.0.
.PP
12 bit samples are not currently supported. This may be fixed if someone
tells me the format used for 12 bit samples - my sound card doesn't
support 12 bit mode. I'm not sure that the sound driver supports them
either though.
.PP
This program runs setuid root. This is required because it uses
setpriority() to run at the highest possible priority, and also locks
down the buffers it uses to avoid them being swapped out.
.SH AUTHOR
David Monro
davidm@cs.usyd.edu.au
.PP
The option parsing code was originally taken from
.B vplay
to maintian
compatability.


Docs/ChangeLog
GramoFile Change Log

version 1.6:

- Included semun definition that disappeared from glibc2.1 headers. Bart
   Warmerdam was the first to submit this to the Debian Bug Tracking system.
- Allow piping the signal processing's output to any command (MP3
   encoders!), thanks to Geoff Clare. See README for details.
- Settings for filtering are now remembered within one run, which makes
   it much easier to try slight variations in the settings. Patch by
   James Tappin.
- New convert-to-mono (=average-channels) filter, also by James Tappin.
- James Tappin also provided an alternative way of interpolation for the
   CMF-II, using an L1 norm linear fit. It's much slower; you can compile
   it in by defining FANCY_FILL in the beginning of signpr_cmf2.c.
- Reverted curses to ncurses per default, should now compile fine on SuSE.
   For the old curses, change $LIBS in the Makefiles and use -DOLD_CURSES.
- Keith Refson sent a scandir() replacement function for Solaris; see the
   README for info on how to use it.
- For byte-swapped systems, the length-fields in the .wav header are
   now byte-swapped too; patch also by Keith Refson.
- FreeBSD patches by Juergen Lock.
- Applied fix for a buffering deadlock at the end of recording, also by
   Juergen Lock.
- James Tappin ported Dave Platt's out-of-phase tick detection (see v1.3)
   to the `old' CMF filter; search for TEST_DAVE_PLATT in signpr_cmf.c.
- Fixed bug in the parameter screen of the double median filter.

version 1.5:

- Signal Processing 30-50% faster, thanks to a few "old programmers'
   tricks" provided by Dave Platt.
- Fix for computation of file offsets and -lengths for long files in
   bplay, and 12 bit/sample should now work correctly, too. Patch by Ralf
   Schlatterbeck.
- Should compile & run on IRIX (and maybe others), too. Joshua Weage sent
   me a patch, which I interpreted a little more generally. More info in
   the README file.

version 1.4:

- `Determining of track separation points' is now called simply
   `Track location'.
- Totally new track location algorithm that should work well in far more
   cases than you're used to ;-)  Read more in Tracksplit2.txt.
- Should compile OK with egcs; patch by Daniel Kobras.

version 1.3:

- Considerable speed improvement in Signal Processing (and Track Splitting,
   too) by using the stdio functions, patch provided by Axel Kohlmeyer.
   (Why didn't anyone tell me speed(fread)>>speed(read) before..?)
- Deleting filters in Filter Selection screen now also works on the console
   (using the R key), again thanks to Axel Kohlmeyer. (I'm always in X ;)
- Info-screen after recording, great idea of Richard Robinson.
- Dave Platt reported that most ticks are out-of-phase, so tick detection
   should look at the left-minus-right (difference) signal. I've programmed
   this in signpr_cmf2.c (search for TEST_DAVE_PLATT) but not tested it;
   it's not used by default. If you like, you may experiment with it - and
   tell me your findings.

version 1.2:

- Added variable median (interpolation) length to the Conditional
   Median Filter, and called it CMF II, so the original is still
   there. The CMF II gives much better results than the original CMF.
- Added the possibility of playing whole tracks, or a few seconds at
   the beginning/end of a track or before/after a track - with a single
   keystroke. Makes life a lot easier.
- Added a Copy Only `filter' that only copies the signal. Useful for
   splitting tracks without processing the signal. Requested (more or
   less) by Richard Robinson.
- Added a Experimenting Filter with which you can experiment. See the
   file signpr_exper.c
- Corrected writing of .wav header - bytes_per_sample seems to be 4
   instead of 2 (as found in bplay.c). Thanks to Pavel Kasparek for
   reporting it.
- Several other small changes.

version 1.1:

- Track splitting operational
- Documented signal processing and track splitting
- English user interface
- Removed some annoying bugs
- Cleaned up source code and used indent(1)
- Webpage at http://cardit.et.tudelft.nl/~card06
- Done a lot of other things

version 1.0:

- First `official' non-public release, at the end of the `Integraal Project
   Practicum', without track splitting, without full documentation (.txt
   files) and a Dutch user interface. 



Source/bplaysrc/bplay.c
/*
** brec/bplay.c (C) David Monro 1996
**
** Copyright under the GPL - see the file COPYING in this directory
**
** Adapted by J.A. Bezemer for use with GramoFile - July 1998
*/

#include <stdio.h>
#include <unistd.h>
#include <errno.h>
#include <string.h>
#include <stdlib.h>
#include <sys/types.h>
#include <fcntl.h>
#include <unistd.h>
#include <sys/stat.h>
#include <sys/wait.h>

#include <sys/time.h>
#include <sys/resource.h>

#ifndef __FreeBSD__
#include <sys/soundcard.h>
#else
#include <machine/soundcard.h>
#endif

#include "fmtheaders.h"

#include "../yesnowindow.h"
void init_curses(void);
void finish_curses(int sig);
#include "../errorwindow.h"
#include "../clrscr.h"
#include "../reclp_main.h"
#ifndef OLD_CURSES
#include <ncurses.h>
#else
#include <curses.h>
#endif

/* Types and constants */
typedef enum sndf_t {F_UNKNOWN, F_WAV, F_VOC, F_RAW} sndf_t;

#define MSPEED	1
#define MBITS	2
#define MSTEREO	4

/* Globals */
char *progname;
int forked;

/* Prototypes */

#ifdef linux
/* This is in libc, but not in the header files. -- but it IS in
   Red Hat 5.0.... Libc6?
   well i'd guess its not in the headers because its nonstandard, i.e.
   (probably) only exists on linux...  -nox */
#ifndef REDHAT50
extern char *basename(char *name);
#endif
#endif

void Usage(void);
void ErrDie(char *err);
void Die(char *err);

void getbcount(int speed, int bits, int stereo, long *bcount,
	long timelim, long samplim, long timejmp, long sampjmp, long *bjump);
void playraw(int thefd, char hd_buf[20], int speed, int bits, int stereo);
void playwav(int thefd, char hd_buf[20], int mods, int speed, int bits,
  int stereo, long timelim, long samplim, long timejmp, long sampjmp );
/* was: void playwav(int thefd, char hd_buf[20], int mods, int speed, int bits, int stereo); */

void playvoc(int thefd, char hd_buf[20]);

/* extern globals */
extern int audio, abuf_size, fmt_mask;

/* extern prototypes */
extern void init_sound(int recorder);
extern void snd_parm(int speed, int bits, int stereo);
extern void init_shm(void);
extern void shmrec(int outfd, long bcount, int terminate);
extern void diskread(int outfd, long bcount, long skipped, char hd_buf[20], 
    int terminate, int speed, int bits, int stereo);
extern volatile void audiowrite(void);
extern void initsems(int disks, int snds);
extern void cleanupsems(void);


void main(int argc, char *argv[])
{
	
	int recorder = 0;		/* 1 if recording, 0 else */
	int thefd;			/* The file descriptor */
	int speed, bits, stereo;	/* Audio parameters */
	long timelim;			/* Recording time in secs */
	long samplim;			/* Recording time in samples */
	long timejmp;			/* Skip time in secs */
	long sampjmp;			/* Skip time in samples */
	long bcount;			/* Number of bytes to record */
	long bjump;			/* Number of bytes to skip */
	int themask;			/* Permission mask for file */
	sndf_t filetype;		/* The file type */
	int mods;			/* So user can override */
	int optc;			/* For getopt */

	init_curses();

#ifdef linux
	progname = basename(argv[0]);	/* For errors */
#else
	progname = strrchr(argv[0], '/');  /* Replacement for e.g. FreeBSD */
	if (!progname || !*++progname)
	    progname = argv[0];
#endif

	/* Ok, find out if we record or play */
	if (strcmp(progname, "brec_gramo") == 0)
	    recorder = 1;
	else
	{
	    if (!(strcmp(progname, "bplay_gramo") == 0))
		Die("must be called as bplay_gramo or brec_gramo");
	    else
		recorder = 0;
	}
	/* Default values */
	speed = 8000; bits = 8; stereo = 0;
	timelim = 0; samplim = 0; bcount = 0;
	timejmp = 0; sampjmp = 0; bjump = 0;
	filetype = F_UNKNOWN;
	mods = 0;
	/* Parse the options */
	while ((optc = getopt(argc, argv, "Ss:b:t:T:j:J:rvw")) != -1)
	{
		switch(optc)
		{
		case 's':
			speed = atoi(optarg);
			if (speed < 300)
				speed *= 1000;
			mods |= MSPEED;
			break;
		case 'b':
			bits = atoi(optarg);
			if ((bits != 8) && (bits != 16))
				Usage();
			mods |= MBITS;
			break;
		case  'S':
			stereo = 1;
			mods |= MSTEREO;
			break;
		case 't':
			timelim = atol(optarg);
			break;
		case 'T':
			samplim = atol(optarg);
			break;
		case 'j':
			timejmp = atol(optarg);
			break;
		case 'J':
			sampjmp = atol(optarg);
			break;
		case 'r':
			filetype = F_RAW;
			break;
		case 'v':
			filetype = F_VOC;
			break;
		case 'w':
			filetype = F_WAV;
			break;
		default:
			Usage();
		}
	}

#if 1
	/* This program is either set-uid or run by root (I hope...) */
	if (setpriority(PRIO_PROCESS, 0, -20) == -1)
#ifndef LP2CD
		fprintf(stderr, "%s: setpriority: %s: continuing anyway\n",
			progname, strerror(errno))
#endif
		;
	
#endif /* 1 */
	/* Drop out of suid mode before we open any files */
	if(setreuid(geteuid(), getuid()) == -1)
	{
#ifndef LP2CD
		fprintf(stderr, "%s: setreuid: %s: continuing anyway\n",
			progname, strerror(errno));
		fprintf(stderr, "real uid = %d, effective uid = %d\n",
			getuid(), geteuid());
#endif
	}

	/* Calculate the time limit in samples if need be */
	if (optind > argc - 1)
	{
	    if(recorder)	/* No file, so stdin or stdout */
		thefd = 1;
	    else
		thefd = 0;
	}
	else
	{
	    /* Open a file */
	    if(recorder)
	    {
		/* Ok, get the mask for the opening the file */
		themask = umask(0);
		umask(themask);
		if ((thefd = open(argv[optind], O_CREAT | O_TRUNC | O_WRONLY,
		    (~themask) & 0666)) == -1)
		    ErrDie(argv[optind]);
	    }
	    else
		if ((thefd = open(argv[optind], O_RDONLY)) == -1)
		    ErrDie(argv[optind]);
	}

	/* Open and set up the audio device */
	init_sound(recorder);

	/* Check if the card is capable of the requested operation */
	/*
	** Can't check for stereo yet, just number of bits. Also,
	** can't check for 12 bit operations yet.
	*/

#if 0
	if ((bits == 8) & !(fmt_mask & AFMT_U8)) Die("Format not supported by audio device");
	if ((bits == 16) & !(fmt_mask & AFMT_S16_BE)) Die("Format not supported by audio device");
#endif

	/* Set up the shared buffers and semaphore blocks */
	init_shm();

	/* Call the appropriate routine */
	if (recorder)
	{
		getbcount(speed, bits, stereo, &bcount,
			 timelim, samplim, timejmp, sampjmp, &bjump);
#ifdef DEBUG
		fprintf(stderr, "bcount: %ld\n", bcount);
#endif


/* By CS: */
			clearscreen(RECLP_HEADERTEXT);
			if (!yesno_window(
"Press Enter to start recording.",
				     " Start ", " Cancel ", 1))
			{
			    close(thefd);
			    close(audio);
			    finish_curses(0);
			}

			header(RECLP_HEADERTEXT);
			error_window_display(
"Press Enter to stop recording.", " Stop ");

/* End CS */


		if (filetype == F_UNKNOWN)
			filetype = F_RAW;	/* Change to change default */
		switch(filetype)
		{
		case F_WAV:
			/* Spit out header here... */
#ifndef LP2CD
			fprintf(stderr, "Writing MS WAV sound file");
#endif
			{
				wavhead header;

				char *riff = "RIFF";
				char *wave = "WAVE";
				char *fmt = "fmt ";
				char *data = "data";

				memcpy(&(header.main_chunk), riff, 4);
				header.length = sizeof(wavhead) - 8 + bcount;
				memcpy(&(header.chunk_type), wave, 4);

				memcpy(&(header.sub_chunk), fmt, 4);
				header.sc_len = 16;
				header.format = 1;
				header.modus = stereo + 1;
				header.sample_fq = speed;
				header.byte_p_sec = ((bits > 8)? 2:1)*(stereo+1)*speed;
/* Correction by J.A. Bezemer: */
				header.byte_p_spl = ((bits > 8)? 2:1)*(stereo+1);
			/* was: header.byte_p_spl = (bits > 8)? 2:1; */

				header.bit_p_spl = bits;

				memcpy(&(header.data_chunk), data, 4);
				header.data_length = bcount;
				write(thefd, &header, sizeof(header));
			}
		case F_RAW:
			if (filetype == F_RAW)
#ifndef LP2CD
				fprintf(stderr, "Writing raw sound file")
#endif
				;
#ifndef LP2CD
			fprintf(stderr, ", %dHz, %dbit, %s\n", speed, bits, (stereo)? "stereo":"");
#endif
			snd_parm(speed, bits, stereo);
			initsems(0, 1);

			shmrec(thefd, bcount, 1);
			break;
		case F_VOC:
			/* Spit out header here... */
			fprintf(stderr, "Writing CL VOC sound file");
			fprintf(stderr, ", %dHz, %dbit, %s\n", speed, bits, (stereo)? "stereo":"");
			{
				vochead header;
				blockTC ablk;
				blockT9 bblk;
				int i;
				char fill = 0;

				for (i=0;i<20;i++)
					header.Magic[i] = VOC_MAGIC[i];
				header.BlockOffset = 0x1a;
				header.Version = 0x0114;
				header.IDCode = 0x111F;
				write(thefd, &header, sizeof(vochead));

				snd_parm(speed, bits, stereo);
				initsems(0, 1);

				i = (bcount >= 0xFFFFF2)? 0xFFFFF2 + 12 : bcount;
				ablk.BlockID = 9;
				ablk.BlockLen[0] = (i + 12) & 0xFF;
				ablk.BlockLen[1] = ((i + 12) >> 8) & 0xFF;
				ablk.BlockLen[2] = ((i + 12) >> 16) & 0xFF;
				bblk.SamplesPerSec = speed;
				bblk.BitsPerSample = bits;
				bblk.Channels = stereo + 1;
				bblk.Format = (bits == 8)? 0 : 4;
				write(thefd, &ablk, sizeof(ablk));
				write(thefd, &bblk, sizeof(bblk));
				shmrec(thefd, i, 1);
				write(thefd, &fill, 1);
			}
			break;
		default:
			Die("internal error - fell out of switch");
		}
	}
	else
	{
		int count;
		pid_t pid;
		char hd_buf[20];	/* Holds first 20 bytes */

		count = read(thefd, hd_buf, 20);
		if (count < 0) ErrDie("read");
		if (count < 20) Die("input file less than 20 bytes long.");

		initsems(1, 0);

		pid = fork();
		if(!pid)
			audiowrite();	/* Doesn't return */
		forked = 1;

	    if (filetype == F_RAW)
			playraw(thefd, hd_buf, speed, bits, stereo);
	    else
	    {
		/* Pick the write output routine */
		if(strstr(hd_buf, VOC_MAGIC) != NULL)
			playvoc(thefd, hd_buf);
		else if(strstr(hd_buf, "RIFF") != NULL)
			playwav(thefd, hd_buf, mods, speed, bits, stereo,
				timelim, samplim, timejmp, sampjmp);
		else /* Assume raw data */
			playraw(thefd, hd_buf, speed, bits, stereo);
	    }
		wait(NULL);
		cleanupsems();
	}

	finish_curses(0);
}

void Usage(void)
{
int i;
	printw(
		"Usage: %s [-S] [-s Hz] [-b 8|16] [-t secs] [-r|-v|-w] [filename]\n",
		progname);
	printw("\nPress any key...");
	refresh();
	i = getch();

	finish_curses(-1);
	exit(1);
}

void ErrDie(char * err)
{
char string[500];
	sprintf(string, "%s: %s", err, strerror(errno));
	error_window(string);
	finish_curses(-1);
	exit(-1);
}

void Die(char * err)
{
char string[500];
	sprintf(string, "%s", err);
	error_window(string);
	finish_curses(-1);
	exit(-1);
}

void getbcount(int speed, int bits, int stereo, long *bcount,
	long timelim, long samplim, long timejmp, long sampjmp, long *bjump)
{
        if(timelim)
        {
/* fprintf(stderr,"tl=%ld, spd=%d, bits=%d\n",timelim, speed, bits); */
                *bcount = speed*timelim*((bits+7)/8L);
/* fprintf(stderr,"bc=%ld\n",bcount); */
                if (stereo) *bcount <<= 1;
/* fprintf(stderr,"bc2=%ld\n",bcount); */
        }
        if(samplim)
        {
                *bcount = samplim*((bits+7)/8L);
                if (stereo) *bcount <<= 1;
        }
        if(timejmp)
        {
                *bjump = speed*timejmp*((bits+7)/8L);
                if (stereo) *bjump <<= 1;
        }
        if(sampjmp)
        {
                *bjump = sampjmp*((bits+7)/8L);
                if (stereo) *bjump <<= 1;
        }
}

void playraw(int thefd, char hd_buf[20], int speed, int bits, int stereo)
{
    fprintf(stderr, "Playing raw data : %d bit, Speed %d %s ...\n",
	bits, speed, (stereo)? "Stereo" : "Mono");
    diskread(thefd, 0,0, hd_buf, 1, speed, bits, stereo);
}

void playwav(int thefd, char hd_buf[20], int mods, int speed, int bits,
  int stereo, long timelim, long samplim, long timejmp, long sampjmp )
{
    wavhead wavhd;
    int count;
long bcount = 0, bjump = 0;

    memcpy((void*)&wavhd, (void*)hd_buf, 20);
    count = read(thefd, ((char*)&wavhd)+20, sizeof(wavhd) - 20);
    if(wavhd.format != 1) Die("Input is not a PCM WAV file");
#ifndef LP2CD
    if (! (mods&MSPEED))
/* This should _not_ be overridden by the commandline options, for those
   are always 44100 Hz 16 bit Stereo, and we want _all_ wavs to
   play correctly. */
#endif
	speed = wavhd.sample_fq;
#ifndef LP2CD
    if (! (mods&MBITS))
#endif
	bits = wavhd.bit_p_spl;
#ifndef LP2CD
    if (! (mods&MSTEREO))
#endif
	stereo = wavhd.modus - 1;
#ifndef LP2CD
    fprintf(stderr, "Playing WAVE : %d bit, Speed %d %s ...\n",
	bits, speed, (stereo)? "Stereo" : "Mono");
#endif
/* By CS: */
    getbcount(speed, bits, stereo, &bcount, timelim, samplim,
		timejmp, sampjmp, &bjump);
    if (bjump) lseek(thefd, bjump, SEEK_CUR);

    diskread(thefd, bcount, bjump, NULL, 1, speed, bits, stereo);
}

void playvoc(int thefd, char hd_buf[20])
{
    int count;
    int speed=0, bits=0, stereo=0;
    int inloop=0, loop_times;
    long bytecount, loop_pos=0;
    vochead vochd;
    blockTC ccblock;
    int lastblocktype = -1;
    int quit = 0;

    fprintf(stderr, "Playing Creative Labs Voice file ...\n");
    memcpy((void*)&vochd, (void*)hd_buf, 20);
    count = read(thefd, ((char*)&vochd)+20, sizeof(vochd) - 20);
    fprintf(stderr, "Format version %d.%d\n", vochd.Version>>8,
	vochd.Version&0xFF);
    if (vochd.IDCode != (~vochd.Version+0x1234))
	fprintf(stderr, "Odd - version mismatch - %d != %d\n",
	    vochd.IDCode, ~vochd.Version+0x1234);
    if(sizeof(vochd) < vochd.BlockOffset)
    {
	int off = vochd.BlockOffset - sizeof(vochd);
	char *junk;
	junk = (char*) malloc(off);    
	read(thefd, junk, off);
    }
    while(!quit)
    {
    if ((read(thefd, (char*)&ccblock, sizeof(ccblock))) == -1)
    {
#ifdef DEBUG
	fprintf(stderr, "Terminating\n");
#endif
	diskread(thefd, -1,0, NULL, 1, speed, bits, stereo);
	quit = 1;
	continue;
    }
#ifdef DEBUG
    fprintf(stderr, "Block of type %d found\n", ccblock.BlockID);
#endif
    switch(ccblock.BlockID)
    {
    case 1:
	{
	blockT1 tblock;
	read(thefd, (char*)&tblock, sizeof(tblock));
	if(tblock.PackMethod != 0) Die("Non PCM VOC block");
	if (lastblocktype != 8)
	{
	    speed = 256000000/(65536 - (tblock.TimeConstant << 8));
	    bits = 8;
	    stereo = 0;
	}
	bytecount = DATALEN(ccblock) -2;
	diskread(thefd, bytecount,0, NULL, 0, speed, bits, stereo);
	lastblocktype = 1;
	}
	break;
    case 8:
	{
	blockT8 tblock;
	read(thefd, (char*)&tblock, sizeof(tblock));
	if(tblock.PackMethod != 0) Die("Non PCM VOC block");
	speed = 256000000/(65536 - tblock.TimeConstant);
	bits = 8;
	stereo = tblock.VoiceMode;
	if (stereo) speed >>=1;
	lastblocktype = 8;
	}
	break;
    case 9:
	{
	blockT9 tblock;
	read(thefd, (char*)&tblock, sizeof(tblock));
	if(tblock.Format != 0 && tblock.Format != 4)
	    Die("Non PCM VOC block");
	speed = tblock.SamplesPerSec;
	bits = tblock.BitsPerSample;
	stereo = tblock.Channels - 1;
	bytecount = DATALEN(ccblock) - 12;
	diskread(thefd, bytecount,0, NULL, 0, speed, bits, stereo);
	lastblocktype = 9;
	}
	break;
    case 0:
#ifdef DEBUG
	fprintf(stderr, "Terminating\n");
#endif
	diskread(thefd, -1,0, NULL, 1, speed, bits, stereo);
	quit = 1;
	break;
    case 6:
	inloop = 1;
	read(thefd, (char*)&loop_times, 2);
	loop_times++;
#ifdef DEBUG
fprintf(stderr, "Beginning loop %d\n", loop_times);
#endif
	loop_pos = lseek(thefd, 0, SEEK_CUR);
	if(loop_pos == -1)
	{
	    fprintf(stderr, "Input not seekable - loop will only be played once\n");
	    loop_times = 1;
	}
	lastblocktype = ccblock.BlockID;
	break;
    case 7:
	if(!inloop)
	{
	    fprintf(stderr, "Loop end with no loop start - ignoring\n");
	    break;
	}
	if(loop_times != 0xFFFF) --loop_times;
	if(loop_times)
	{
#ifdef DEBUG
	    fprintf(stderr, "Looping...\n");
#endif
	    lseek(thefd, loop_pos, SEEK_SET);
	}
	else
	    inloop = 0;
	lastblocktype = ccblock.BlockID;
	break;
    default:
	{
	int rd = 0, trgt = BUFSIZ;
	char junkbuf[BUFSIZ];
    
	fprintf(stderr, "Ignored\n");
	bytecount = DATALEN(ccblock);
	while(rd < bytecount)
	{
	    if (rd + trgt > bytecount)
		trgt = bytecount - rd;
	    count = read(thefd, junkbuf, trgt);
	    if (count < 0) ErrDie("read");
	    if (count == 0) Die("premature eof in input");
	    rd += count;
	}
	lastblocktype = ccblock.BlockID;
	}
	break;
    }
    }
}

/* ----- BELOW: ADDED BY CS ------------------------------------------------ */


void init_curses(void)
{
	initscr();
	keypad(stdscr, TRUE);
	nonl();
	cbreak();
	noecho();

	refresh(); /* Xterms erase everything after the first refresh,
		      so refresh one time before anything is added. */

	return;
}

void finish_curses(int sig)
{
	endwin();
	exit(0);
}

